rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,394 @@
+# RFC 12217 - 2022-08-25 - Handling Discarded Events
+
+We want to give Vector users the assurance that events are never lost unexpectedly. There are
+currently many circumstances in which events are simply discarded without recourse for users of
+Vector to recover those events. This RFC presents a framework for handling those discarded events in
+a uniform fashion in all components.
+
+## Context
+
+- [Support dead letter queue sources](https://github.com/vectordotdev/vector/issues/1772)
+- [Vector sources acknowledge discarded events](https://github.com/vectordotdev/vector/issues/12217)
+- [Reroute discarded events from throttle transform](https://github.com/vectordotdev/vector/issues/13549)
+- Sink Error Handling RFC (internal document)
+
+## Cross cutting concerns
+
+- The proposed changes will impact emission and receipt of end-to-end acknowledgements.
+- The proposed configuration options should fit in the configuration schema, but may have
+  implications for its interpretation, particularly around the topology connections.
+- Handling discarded events in any way will change the interpretation of several exiting internal
+  metrics.
+
+## Scope
+
+### In scope
+
+- Handling of discarded events in sources, transforms, and sinks.
+
+### Out of scope
+
+- Handling of discarded data in sources before it is translated into events.
+- Handling of discarded sink request data after it is translated from events.
+
+## Pain
+
+Vector has a number of ways in which components may drop events due to errors that are outside of
+the operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+
+This leads to the potential for lost events and even silently dropped events. Users want to be able
+to capture discarded events at various points in the topology in order to forward them to fail-safe
+destinations for diagnosis.
```

Comment:
or even just to extract metrics from them to increase observability (for example, we provide a vector-based service to our customers and would like to be able to tell how many events were discarded per customer)

Reply:
We already emit telemetry to track dropped events (i.e. `component_discarded_events_total`), whereas the above is merely referring to a new mechanism to actually allow trying to _do_ something with those events before deciding to drop them entirely.